### ISSUE TITLE
startJVM fails without any message with Python 3.8 32bit and jre1.8.0_241 on Windows 10

### ISSUE THREAD

**Issue by Matthias1975** [jpype-project/jpype#672]
It seems, that it fails in jp_exception.cpp (line 493):

![image](https://user-images.githubusercontent.com/63334343/78748762-1a6d7a80-796d-11ea-8c01-57ac4384c852.png)

There is no exception caught in the catch-Block at jp_env.cpp (line 184):

![image](https://user-images.githubusercontent.com/63334343/78748810-307b3b00-796d-11ea-8cd3-5d3e9fd0f818.png)

jpype1 was buildt with:
![image](https://user-images.githubusercontent.com/63334343/78749123-e8104d00-796d-11ea-9e45-530d65eb82ab.png)

the result is:
![image](https://user-images.githubusercontent.com/63334343/78749144-f5c5d280-796d-11ea-8a6a-24098868e9b3.png)

What could be the reason?

**Comment by Matthias1975** [comment]
Windows Event Viewer shows:
Name der fehlerhaften Anwendung: python.exe, Version: 3.8.2150.1013, Zeitstempel: 0x5e55a3ed
Name des fehlerhaften Moduls: _jpype.cp38-win32.pyd, Version: 0.0.0.0, Zeitstempel: 0x5e8d6107
**Exception Code: 0xc0000409
Error offset: 0x0001696d**
ID des fehlerhaften Prozesses: 0x5694
Startzeit der fehlerhaften Anwendung: 0x01d60d69a19940cf
Pfad der fehlerhaften Anwendung: c:\DEVTOOLS\Python\38-32\python.exe
Pfad des fehlerhaften Moduls: c:\DEVTOOLS\Python\38-32\lib\site-packages\jpype1-0.7.2-py3.8-win32.egg\_jpype.cp38-win32.pyd
Berichtskennung: e758c246-05c9-477c-9272-a4715b5a616a
Vollständiger Name des fehlerhaften Pakets: 
Anwendungs-ID, die relativ zum fehlerhaften Paket ist: 

Regarding to the link below 0xc0000409 means STATUS_STACK_BUFFER_OVERRUN.
https://stackoverflow.com/questions/23409809/windows-7-exception-code-0xc0000409

**Comment by Thrameos** [comment]
Ouch that is a bad place to throw an error.   At that particular point we don't yet have a working set of resources to handle an exception.  But the fact that the exception isn't caught is a general indication that something happened during the JNI call.  So why would it not be caught?

Our current exception stack only captures exceptions that are produced by our handling calls except for the rethrow.  So lets walk down the exception handling.

Each frame catches and adds stack info to the exception (which could potentially go wrong, but it would require a serious problem.
```
#define JP_TRACE_OUT } catch (JPypeException &ex) { ex.from(JP_STACKINFO()); throw; }
```

But if it hit this code path,then it should have printed a message prior to getting to the catch.
```
void JPypeException::from(const JPStackInfo& info)
{
	JP_TRACE("EXCEPTION FROM: ", info.getFile(), info.getLine());  // <= if it were JPypeException this get printed
	m_Trace.push_back(info);
}
```

So it must be a different type of exception.  Falling back to the catch for Python, all exceptions are redirected to the exception master handler.
```
#define JP_PY_CATCH(...)  } catch(...) \
  { PyJPModule_rethrow(JP_STACKINFO()); } \
  return __VA_ARGS__
```

That will then transfer all exceptions that are uncaught to the handler.  But there is another alarm bell here.  In order to get transferred it must have printed a message that it is being rethrown. 
```
void PyJPModule_rethrow(const JPStackInfo& info)
{
	JP_TRACE_IN("PyJPModule_rethrow");  //<= But this line has to be printed 
	JP_TRACE(info.getFile(), info.getLine());
	try
	{
		throw;
	} catch (JPypeException& ex) // <= we only handle JPype exceptions
	{
		ex.from(info); // this likely wont be necessary, but for now we will add the entry point.
		ex.toPython();
		return;
	}
        // Nothing handles case of a fail other than JPype.
	JP_TRACE_OUT;  // But if we did get to here then we are now stuck in an endless loop,
}
```

So I do see one way that we could get a stack corruption here.   If we get an unexpected exception type, we will burn through the stack resulting in the error reported.

So how to address this.

First lets add a fatal error handling routine to PyJPModule_rethrow which will address part of the bug.  But the harder part is understanding how we could have gotten an exception from ``frame.FindClass("java/lang/Throwable")``.

**Comment by Thrameos** [comment]
I should mention that this error has occurred before, but I have never successfully replicated it locally.  My general opinion it is some corrupt build.  Pulling the latest Python 32 bit and running against Oracle 1.8 241 works fine on my system for what it is worth.

```
$ python
Python 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 22:45:29) [MSC v.1916 32 bit (In
tel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import jpype
>>> print(jpype.getDefaultJVMPath())
C:\Program Files (x86)\Java\jre1.8.0_241\bin\client\jvm.dll
>>> jpype.startJVM(convertStrings=False)
>>>
```

**Comment by Matthias1975** [comment]
Do you use Win 10 64bit?
I will reinstall Python and retry...

**Comment by Thrameos** [comment]
I use windows 10 and can successfully run under the following configurations.

- Python amd64 between 3.5 and 3.8 with oracle and corretto 8 and 11  (installed from conda)
- Python x86 3.8 with oracle 8 (installed from Python.com)
- Python WSL 3.8 with openjdk 8 and 11 (installed ubuntu)

It appears my production machine is currently "Python 3.7 conda with open JDK 12" on Windows 10.

```
Python 3.7.3 (default, Mar 27 2019, 17:13:21) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32

This Python interpreter is in a conda environment, but the environment has
not been activated.  Libraries may fail to load.  To activate this environment
please see https://conda.io/activation

Type "help", "copyright", "credits" or "license" for more information.
>>> import jpype
>>> print(jpype.getDefaultJVMPath())
C:\Program Files\Java\jdk-12.0.1\bin\server\jvm.dll
>>> jpype.startJVM(convertStrings=False)
>>>
```

So I am very sure there are working configurations.  I just have no clue what could be going wrong on your system.

**Comment by Thrameos** [comment]
See #673 for exception handling fix.

**Comment by Matthias1975** [comment]
I reinstalled Python and jpype (using pip):

C:\DEVTOOLS\Python\38-32>python
Python 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 22:45:29) [MSC v.1916 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import jpype
>>> print(jpype.getDefaultJVMPath())
C:\Program Files (x86)\Java\jre1.8.0_241\bin\client\jvm.dll
>>> jpype.startJVM(convertStrings=False)

C:\DEVTOOLS\Python\38-32>

**Comment by Thrameos** [comment]
Hmm.  Lets try a version with the C++ exception fix

```
 python -m pip install git+https://github.com/Thrameos/jpype.git@exception_bug
```

if you don't mind?

**Comment by Matthias1975** [comment]
pip fails:
![image](https://user-images.githubusercontent.com/63334343/78800965-5da30a00-79bc-11ea-9ff2-a591405ed998.png)

First I uninstalled jpype and then reinstalled it. 

It seems that "JPypeClassLoader.java" is missing.

**Comment by Thrameos** [comment]
Darn a two-for-one.  I will have to think on this a bit as clearly we are getting very different outcomes.   Thank you for the bug report, hopefully I can get this resolved when I am back in. 

```
$ mkdir ~/tmp
$ cd ~/tmp
$ python -m pip install git+https://github.com/Thrameos/jpype.git@exception_bug
Collecting git+https://github.com/Thrameos/jpype.git@exception_bug
  Cloning https://github.com/Thrameos/jpype.git (to revision exception_bug) to c
:\cygwin64\tmp\pip-req-build-l1o3eqwm
  Running command git clone -q https://github.com/Thrameos/jpype.git 'C:\cygwin6
4\tmp\pip-req-build-l1o3eqwm'
  Running command git checkout -b exception_bug --track origin/exception_bug
  Switched to a new branch 'exception_bug'
  Branch 'exception_bug' set up to track remote branch 'exception_bug' from 'ori
gin'.
Building wheels for collected packages: JPype1
  Building wheel for JPype1 (setup.py) ... done
  Created wheel for JPype1: filename=JPype1-0.7.2-cp38-cp38-win32.whl size=12533
55 sha256=7ca33c179478b4bd69caaa8e4d53cfe19774e49105afb3ca2d07070308be7f8b
  Stored in directory: C:\cygwin64\tmp\pip-ephem-wheel-cache-0a_g1gz5\wheels\a1\
f1\2f\93ce2498590216c151eeea9ccaeaea65ee38722f8dc5003037
Successfully built JPype1
Installing collected packages: JPype1
Successfully installed JPype1-0.7.2
(python-3.8-win32)
$
```

**Comment by Thrameos** [comment]
My German is not so great, but could that mean that "javac" is not found rather than "JPypeClassLoader.java"?  It would give a different error if it could not find the java file.

https://www.easeus.de/daten-wiederherstellen/das-system-kann-die-angegebene-datei-nicht-finden.html

**Comment by Matthias1975** [comment]
Obviously your German is better than mine ;)
Adding java to %PATH% made it work.

The behavior is still the same though
![image](https://user-images.githubusercontent.com/63334343/78805749-191a6d00-79c2-11ea-949f-8e1c700314e1.png)

**Comment by Thrameos** [comment]
No errors, no output, nada.  I am stumped.

At this point I would recommend trying to capture it under a debugger or using the trace with some printfs thrown in.   The debugger will be  a problem as the JVM uses segfault as a messaging system for the gc so first thing it does is create an error.  So I guess printf is your best bet.

I will be whatever assistance that I can be.

**Comment by Matthias1975** [comment]
Thank you so far.

I will spend some time on it...

**Comment by Matthias1975** [comment]
I got the master and building it with "--enable-tracing" fails, with compiler errors:
- jp_class (343):  `JP_TRACE("Python");//, JPPyObject::getTypeName(pyobj));` <- `pyobj` not defined
- jp_exception.cpp(30): `JP_TRACE("JAVA EXCEPTION THROWN:");//, msn);` <- `msn` not defined
- jp_exception.cpp(40): `JP_TRACE("EXCEPTION THROWN:", error);//, msn);` <- `msn` not defined
- jp_proxy.cpp(47):
`// JP_TRACE("Convert", i, type->getCanonicalName());` <- `type` not defined yet
`jobject obj = frame.GetObjectArrayElement(args, i);`
`JPClass* type = frame.findClassForObject(obj);`
`JP_TRACE("Convert", i, type->getCanonicalName());`  <- line moved
- pyjp.h (40/41):
the macro `JP_PY_CATCH_NONE` is not defined, with `JP_TRACING_ENABLE` set
`#endif`
`#define JP_PY_CATCH_NONE(...)  } catch(...) {} return __VA_ARGS__` <- switched both lines

I would suggest to continue when these errors are fixed.

**Comment by Thrameos** [comment]
Fixed in #675

**Comment by Matthias1975** [comment]
Is it possible there is something wrong with JPJavaFrame?

These lines seem to work:
![image](https://user-images.githubusercontent.com/63334343/78972859-06fa1500-7b0f-11ea-8687-8610588fb386.png)

But the similar code fails:
![image](https://user-images.githubusercontent.com/63334343/78972878-14af9a80-7b0f-11ea-86af-d04cd405f7ba.png)

Result:
![image](https://user-images.githubusercontent.com/63334343/78972909-209b5c80-7b0f-11ea-9606-e10ac85e55e0.png)

**Comment by Matthias1975** [comment]
@Karl: 
is it possible to get your compiled pyd-files to try them on my machine? 
Maybe the compiler makes a difference...?

**Comment by Thrameos** [comment]
Okay so we have it isolated to one statement.  We just need to break it down a bit further to see what happened.

```
jmethodID JPJavaFrame::GetMethodID(jclass a0, const char* a1, const char* a2)
{
JP_TRACE_IN("GetMethodID", this);
	JAVA_CALL("JPJavaFrame::GetMethodID");  // This is setting up exception handling.
JP_TRACE("env", m_Env);
JP_TRACE("a1", a1);
JP_TRACE("a2", a2);
try{
	jmethodID mid = m_Env->functions->GetMethodID(m_Env, a0, a1, a2);  // This is a native call the JVM DLL
JP_TRACE("GetMethodID success");
       return mid;
catch(std::exception &ex)
{
JP_TRACE("unexpected throw", ex.what());
}
catch(...)
{
JP_TRACE("unexpected throw");
}
JP_TRACE_OUT;
}
```

From here we need to see where that JAVA_CALL statement went.
```
#ifdef JP_INSTRUMENTATION
#define JAVA_CALL(X) \
	JPCall call(*this); \
	PyJPModuleFault_throw(compile_hash(X))
#else
#define JAVA_CALL(X) JPCall call(*this);
#endif
#define JAVA_CHECK(X)  call.check()
```

So here is the most dangerous part of this code path
```
/** Call a java jni function with error checking.
 *
 * This version should be used when the java call is a fixed length of time
 * to complete.
 */
class JPCall
{
	JNIEnv* _env;
	JPContext* _context;
public:

	JPCall(const JPJavaFrame& frame)
	{
// We are just setting up some variables for later.
		_env = frame.getEnv();
		_context = frame.getContext();
	}

	void check()
	{
		if (_env && _env->functions->ExceptionCheck(_env) == JNI_TRUE)
		{
			JPJavaFrame frame(_context, _env);
			jthrowable th = _env->functions->ExceptionOccurred(_env);
			_env->functions->ExceptionClear(_env);
			_env = 0;
			throw JPypeException(frame, th, JP_STACKINFO());
		}
	}

// This is a throw from a dtor.  Which is intended to make sure that we check for JNI errors....  
// But if somehow we enter this from a C++ exception we would be in trouble.  
	~JPCall() NO_EXCEPT_FALSE  
	{
// So we need to instrument it
JP_TRACE_IN("~JPCall");
JP_TRACE("Check exception", _env);
		// This is a throw in destructor which is only allowed on an exception safe code pattern
		if (_env != NULL && _env->functions->ExceptionCheck(_env) == JNI_TRUE)
		{
JP_TRACE("Set up frame", _context, _env); 
			JPJavaFrame frame(_context, _env);
JP_TRACE("Check exception");
			jthrowable th = _env->functions->ExceptionOccurred(_env);
JP_TRACE("Clear exception", th);
			_env->functions->ExceptionClear(_env);
			_env = 0;
JP_TRACE("Transfer control to throw"); // next line can abort if we have something already thrown
			throw JPypeException(frame, th, JP_STACKINFO());  // lgtm [cpp/throw-in-destructor]
		}
JP_TRACE_OUT;
	}
} ;
```

So the three likely problems we have to rule out
- [ ] your JVM is not acting properly, (did it die in the native call?) 
- [ ] your compiler failed to respect the no-exception line and it barfing on the check and has compiled into invalid code.  (did it die in the exception check?)
- [ ] is there some way to get another exception issues.

In addition we have another alternative approach, which is trying to determine if it is something special about the call to "getStackTrace".   In order to rule that out I would attempt to take some minor piece of code like getting the method for toString above the crash point.

**Comment by Matthias1975** [comment]
Here the output:
````
>>> jpype.startJVM(convertStrings=False)
<B msg="PyJPModule_startup" >
  <B msg="JPPyUnicode::asStringUTF8" >
    <M>JPPyUnicode::asStringUTF8 : pyref new(claim) 026FDB60 1 bytes</M>
    <M>JPPyUnicode::asStringUTF8 : pyref dtor(dec) 026FDB60 1 bytes</M>
  </B> <!-- JPPyUnicode::asStringUTF8 -->
  <M>PyJPModule_startup : vmpath C:\Program Files (x86)\Java\jre1.8.0_241\bin\client\jvm.dll</M>
  <M>PyJPModule_startup : pyref new(inc) 00150028 3102 tuple</M>
  <M>PyJPModule_startup : pyref new(claim) 02743870 20 module</M>
  <M>PyJPModule_startup : pyref new(claim) 02767A48 4 list</M>
  <M>PyJPModule_startup : pyref new(claim) 0269FE88 3 builtin_function_or_method</M>
  <M>PyJPModule_startup : pyref dtor(dec) 0269FE88 4 builtin_function_or_method</M>
  <M>PyJPModule_startup : pyref dtor(dec) 02767A48 4 list</M>
  <M>PyJPModule_startup : pyref dtor(dec) 02743870 20 module</M>
  <B msg="JPContext::startJVM" >
    <B msg="JPException_init" >
      <B msg="~JPCall" >
        <M>~JPCall : Check exception 02EED140</M>
      </B> <!-- ~JPCall -->
      <B msg="GetMethodID" >

D:\jpype-master>
````

**Comment by Thrameos** [comment]
That is beyond strange.  There is nothing between the last print and the print of the args that could do anything.   It is just calling a ctor that copies two variables. Most likely the stack has been corrupted before this point.   

I will put up a test PR after work to try to make this code a bit safer.

**Comment by Thrameos** [comment]
I attempted to address a potential problem in #679

**Comment by Thrameos** [comment]
I haven't heard back from you.  I did find one matching bug report indicating that on some systems it is necessary to place the java bin directory in the path to get the JVM to load.

**Comment by Matthias1975** [comment]
I set PATH to 

```
> PATH=C:\Program Files (x86)\Windows Kits\10\bin\10.0.18362.0\x86;C:\DEVTOOLS\Java\1.8-32\bin;...

>>> import jpype
>>> jpype.startJVM(convertStrings=False)
```

and got the same results.

**Comment by Thrameos** [comment]
@Matthias1975   I added a set of files to help you with debugging the start up issue in PR #729.   

The instructions are in project/debug/windows/README.rst.  It researching the problem I found one bug with the error reporting path for JVM, but just prevented it from reporting a fairly meaningless message like "unable to load %1".   Windows simply does not produce any usable diagnostics information on LoadLibrary.

I ran through every possible test case local.  (Missing DLL, corrupt DLL, wrong Arch, missing JVM dependency) and recorded the outputs I got.   Hopefully it will help you find your issue.  The problem is clearly a site problem.  You have an improperly installed JVM, corrupt libraries, or paths pointing to libraries that don't belong to the JVM.  There is nothing further I can do.
